### PR TITLE
Track dependencies in order to invalidate them when an object is invalidated

### LINKED ISSUE #1695: Item action button disabled after redirect from action page
**Describe the bug**
After withdrawing/reinstating an Item and getting redirected back to `/edit/status`, the reinstate/withdraw button is disabled.

**To Reproduce**
1. Log in as an admin
2. Edit any archived Item
3. Click "Withdraw" and confirm
   * You will be redirected back to `/edit/status`
   * The "Reinstate" button will be disabled
4. Refresh
   * The "Reinstate" button will no longer be disabled
5. Click "Reinstate" and confirm
   * You will be redirected back to `/edit/status`
   * The "Withdraw" button will be disabled
6. Refresh
   * The "Withdraw" button will no longer be disabled

**Expected behavior**
The buttons should _not_ be disabled after the initial redirect.

**Underlying issue**
This happens because the Item's authorization features are taken from the cache after the redirect; if the action changes them this does not get reflected on the status page.
Currently this is only noticeable for `withdrawItem` & `reinstateItem`.

@artlowel suggested this should be solved more generally:
* When a cached DSO is invalidated via [`DataService.invalidate()`](https://github.com/DSpace/dspace-angular/blob/26ed23d1b6ea61922de1ee431aba56dc35afd9ca/src/app/core/data/data.service.ts#L583-L583), cached authorizations related to it should be invalidated as well
* We'd need to have a way of retrieving the related authorizations, for example by keeping an index of UUID → authorization.
 Even more generally, we may want to index _resources that depend on other resources_; DSO → authorization would be a specific case of this.
* Once something like the above is in place, the bug can be solved by adding an `invalidate(item)` call to [`ItemDataService.setWithDrawn`](https://github.com/DSpace/dspace-angular/blob/26ed23d1b6ea61922de1ee431aba56dc35afd9ca/src/app/core/data/item-data.service.ts#L132-L132) and similar methods

**Related work**
#1608

### PR DESCRIPTION
**Pull request by ybnd**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #1695
* Builds on #1813

## Description
This PR addresses the bug where Item the actions "Withdraw" & "Reinstate" would remain disabled on `items/.../edit/status` upon redirect after reinstating or withdrawing the Item.

We do this by 

* introducing a mechanism to track dependencies between cached objects
* making the `isAuthorized` search request dependent on the `objectUrl`
* ensuring that cached objects are invalidated on after successful patch requests, so that authorizations can be invalidated when the related Item is invalidated.

## Instructions for Reviewers
Confirm that #1695 can no longer be replicated

## Checklist

_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/core/cache/object-cache.actions.ts [hunk 1]
```diff
@@ -126,13 +128,49 @@ export class ApplyPatchObjectCacheAction implements Action {
   }
 }
 
+export class AddDependentsObjectCacheAction implements Action {
```
Comment (tdonohue) [inline src/app/core/cache/object-cache.actions.ts:131]: Can we describe this action via TypeDocs or comments?

File: src/app/core/cache/object-cache.actions.ts [hunk 2]
```diff
@@ -126,13 +128,49 @@ export class ApplyPatchObjectCacheAction implements Action {
...
+   */
+  constructor(href: string, dependentRequestUUIDs: string[]) {
+    this.payload = {
+      href,
+      dependentRequestUUIDs,
+    };
+  }
+}
+
+export class RemoveDependentsObjectCacheAction implements Action {
```
Comment (tdonohue) [inline src/app/core/cache/object-cache.actions.ts:152]: Same here, it'd be best to describe this via TypeDocs or comments

File: src/app/core/cache/object-cache.actions.ts [hunk 3]
```diff
@@ -126,13 +128,49 @@ export class ApplyPatchObjectCacheAction implements Action {
...
+  type = ObjectCacheActionTypes.REMOVE_DEPENDENTS;
+  payload: string;
+
+  /**
+   * Create a new AddDependencyObjectCacheAction
```
Comment (tdonohue) [inline src/app/core/cache/object-cache.actions.ts:157]: This is a new `RemoveDependentsObjectCacheAction`

File: src/app/core/cache/object-cache.actions.ts [hunk 4]
```diff
@@ -126,13 +128,49 @@ export class ApplyPatchObjectCacheAction implements Action {
...
 
+export class AddDependentsObjectCacheAction implements Action {
+  type = ObjectCacheActionTypes.ADD_DEPENDENTS;
+  payload: {
+    href: string;
+    dependentRequestUUIDs: string[];
+  };
+
+  /**
+   * Create a new AddDependencyObjectCacheAction
```
Comment (tdonohue) [inline src/app/core/cache/object-cache.actions.ts:139]: Typo, this is creating an `AddDependentsObjectCacheAction`

**Review by tdonohue — COMMENTED** [review]
@ybnd : Gave this a try today & this fixes the bug described in #1695!  The code also looks good overall.  I'm basically a +1, but I have a few small requests for TypeDocs below.

**Review by atarix83 — APPROVED** [review]
thanks @ybnd  for this PR

I've verified bug is not present. Code LGTM

**Review by tdonohue — APPROVED** [review]
👍 Thanks for the minor updates, @ybnd .  This looks good to me now too!

**Pull request merged**